geofileops/util/_geoops_sql.py

```diff
@@ -2146,41 +2167,137 @@ def _prepare_filter_by_location_fields(
-        )
-        spatial_relation_filter = f'{subquery_alias}."GFO_$TEMP$_SPATIAL_RELATION" = 1'
-        true_for_disjoint = False
-
-        return (spatial_relation_column, spatial_relation_filter, true_for_disjoint)
-
-    # It is a more complex query, so some more processing needed
-    spatial_relations_filter = _prepare_spatial_relations_filter(query)
-    spatial_relation_column = (
-        ',ST_relate({input1}, {input2}) AS "GFO_$TEMP$_SPATIAL_RELATION"'
+    spatial_relation_column: str = ""
+    spatial_relation_filter: str = ""
+    true_for_disjoint = False
+    groupby = ""
+    spatial_relations = [
+        "disjoint",
+        "equals",
+        "touches",
+        "within",
+        "overlaps",
+        "crosses",
+        "intersects",
+        "contains",
+        "covers",
+        "coveredby",
+    ]
+
+    # Add a specific optimisation for spatial relations.
+    if query.split()[0] == "disjoint":
+        disjoint = True
+        relation_is_true = False
+    else:
+        disjoint = False
+        relation_is_true = query.split()[-1] == "True"
+    groupby = (
+        "GROUP BY layer2.fid_1" if subdivide_coords > 0 and relation_is_true else ""
     )
-    spatial_relation_filter = spatial_relations_filter.format(
-        spatial_relation=f'{subquery_alias}."GFO_$TEMP$_SPATIAL_RELATION"'
+    geom2 = (
+        f"ST_union({geom2})"
+        if subdivide_coords > 0 and relation_is_true
+        else f"{geom2}"
     )
+    for spatial_relation in spatial_relations:
+        if query.lower() == f"{spatial_relation} is {str(relation_is_true).lower()}":
+            geomA = geom2 if spatial_relation == "contains" else geom1
+            geomB = geom1 if spatial_relation == "contains" else geom2
+            spatial_relation_column = (
+                f",ST_{spatial_relation}({geomA}, {geomB})"
+                'AS "GFO_$TEMP$_SPATIAL_RELATION"'
+            )
+            spatial_relation_filter = (
+                f'{subquery_alias}."GFO_$TEMP$_SPATIAL_RELATION" = '
+                f"{int(relation_is_true)}"
+            )
+            true_for_disjoint = not disjoint if not relation_is_true else disjoint
+            break
 
-    # Determine if the spatial_relations_query returns True for disjoint features
-    spatial_relation_column_disjoint = spatial_relation_column.format(
-        input1="ST_GeomFromText('POLYGON((0 0, 0 1, 1 1, 1 0, 0 0))')",
-        input2="ST_GeomFromText('POLYGON((5 0, 5 1, 6 1, 6 0, 5 0))')",
+    return (
+        spatial_relation_column,
+        spatial_relation_filter,
+        true_for_disjoint,
+        groupby,
     )
-    test_path = Path(__file__).resolve().parent / "test.gpkg"
-    sql_stmt = f"""
-        SELECT * FROM (
-            SELECT NULL AS ignore
-                  {spatial_relation_column_disjoint}
-            ) {subquery_alias}
-         WHERE {spatial_relation_filter}
+
+
+def _prepare_filter_by_location_fields(
```

Comment:
Same here, probably useful to add some tests on this function as well for some "special cases".